velox/exec/fuzzer/WindowFuzzer.cpp

```diff
@@ -0,0 +1,299 @@
+  VELOX_CHECK(
+      FLAGS_steps > 0 || FLAGS_duration_sec > 0,
+      "Either --steps or --duration_sec needs to be greater than zero.")
+
+  auto startTime = std::chrono::system_clock::now();
+  size_t iteration = 0;
+
+  while (!isDone(iteration, startTime)) {
+    LOG(INFO) << "==============================> Started iteration "
+              << iteration << " (seed: " << currentSeed_ << ")";
+
+    auto signatureWithStats = pickSignature();
+    signatureWithStats.second.numRuns++;
+
+    auto signature = signatureWithStats.first;
+    stats_.functionNames.insert(signature.name);
+
+    const bool customVerification =
+        customVerificationFunctions_.count(signature.name) != 0;
+    const bool requireSortedInput =
+        orderDependentFunctions_.count(signature.name) != 0;
+
+    std::vector<TypePtr> argTypes = signature.args;
+    std::vector<std::string> argNames = makeNames(argTypes.size());
+
+    auto call = makeFunctionCall(signature.name, argNames, false);
+
+    std::vector<SortingKeyAndOrder> sortingKeysAndOrders;
+    // 50% chance without order-by clause.
+    if (vectorFuzzer_.coinToss(0.5)) {
+      sortingKeysAndOrders =
+          generateSortingKeysAndOrders("s", argNames, argTypes);
+    }
+    auto partitionKeys = generateSortingKeys("p", argNames, argTypes);
+    auto frameClause = generateFrameClause();
+    auto input = generateInputDataWithRowNumber(argNames, argTypes, signature);
+    // If the function is order-dependent, sort all input rows by row_number
+    // additionally.
+    if (requireSortedInput) {
+      sortingKeysAndOrders.push_back(
+          SortingKeyAndOrder("row_number", "asc", "nulls last"));
+      ++stats_.numSortedInputs;
+    }
+
+    logVectors(input);
+
+    bool failed = verifyWindow(
+        partitionKeys,
+        sortingKeysAndOrders,
+        frameClause,
+        call,
+        input,
+        customVerification,
+        FLAGS_enable_window_reference_verification);
+    if (failed) {
+      signatureWithStats.second.numFailed++;
+    }
+
+    LOG(INFO) << "==============================> Done with iteration "
+              << iteration;
+
+    if (persistAndRunOnce_) {
+      LOG(WARNING)
+          << "Iteration succeeded with --persist_and_run_once flag enabled "
+             "(expecting crash failure)";
+      exit(0);
+    }
+
+    reSeed();
+    ++iteration;
+  }
+
+  stats_.print(iteration);
+  printSignatureStats();
+}
+
+void WindowFuzzer::go(const std::string& /*planPath*/) {
+  // TODO: allow running window fuzzer with saved plans and splits.
+  VELOX_NYI();
+}
+
+void WindowFuzzer::updateReferenceQueryStats(
+    AggregationFuzzerBase::ReferenceQueryErrorCode ec) {
+  if (ec == ReferenceQueryErrorCode::kReferenceQueryFail) {
+    ++stats_.numReferenceQueryFailed;
+  } else if (ec == ReferenceQueryErrorCode::kReferenceQueryUnsupported) {
+    ++stats_.numVerificationNotSupported;
+  }
+}
+
+std::string retrieveFunctionName(const core::PlanNodePtr& node) {
```

Comment:
nit: This is usefull for aggfuzzer also , It would be nice if you can move it aggregationBase and have it take the type of node as a type parameter ?

Reply:
I can move it to AggregationFuzzerBase, but making the type of node being parameterized wouldn't be useful. The reason is that this piece of code extract the function name from the typed plan node, while the function name is stored in different structs between AggregationNode and WindowNode.